### PR TITLE
FFT View: Graphical Performance Improvements

### PR DESCRIPTION
**Pull request by Matt54**
FFTView: Graphical Performance Improvements
- Added metal rendering of View
- Removed unnecessary calculations on fftData that are not used in visualization
- Added temporary array of amplitude floats to swap the amplitude array instead of adjusting element by element (I suspected that this results in less SwiftUI update events, but I am not certain)

### CODE REVIEW

**Comment by aure** [comment]
Woohoo!

File: Sources/AudioKitUI/Visualizations/FFTView.swift [hunk 1]
```diff
@@ -22,38 +22,45 @@ class FFTModel: ObservableObject {
...
-                    self.amplitudes[i / 2] = mappedAmplitude
+                // get the real and imaginary parts of the complex number
+                let real = fftData[i]
+                let imaginary = fftData[i + 1]
+                
+                let normalizedBinMagnitude = 2.0 * sqrt(real * real + imaginary * imaginary) / Float(FFT_SIZE)
+                let amplitude = Double(20.0 * log10(normalizedBinMagnitude))
+                
+                // scale the resulting data
+                var scaledAmplitude = (amplitude + 250) / 229.80
```
Comment (eljeff) [inline Sources/AudioKitUI/Visualizations/FFTView.swift:41]: What's the meaning of these numbers (250 & 229.80)? Were these just tweaked by trial and error? Do they work at different sizes?
Could / should they be parameters of the view?
Thanks for doing this.

File: Sources/AudioKitUI/Visualizations/FFTView.swift [hunk 2]
```diff
@@ -22,38 +22,45 @@ class FFTModel: ObservableObject {
...
+                // restrict the range to 0.0 - 1.0
+                if scaledAmplitude < 0 {
+                    scaledAmplitude = 0
+                }
+                if scaledAmplitude > 1.0 {
+                    scaledAmplitude = 1.0
                 }
+                    
+                // further scaling array to look good in visualizer
+                var mappedAmplitude = self.map(n: scaledAmplitude, start1: 0.3, stop1: 0.9, start2: 0.0, stop2: 1.0)
```
Comment (eljeff) [inline Sources/AudioKitUI/Visualizations/FFTView.swift:52]: Same comment for these numbers here, 0.3 & 0.9 - just wondering if they could/should be parameterised or calculated somehow.

**Review by eljeff — COMMENTED** [review]
Mostly just asking questions about 'magic' numbers that I don't immediately understand - this is awesome.

File: Sources/AudioKitUI/Visualizations/FFTView.swift (hunk 1 above)
Comment (Matt54) [reply Sources/AudioKitUI/Visualizations/FFTView.swift:41]: Thanks for checking out the code. I've been wondering the same thing - I wrote this line awhile back so my memory is a bit fuzzy. I do wonder if we can realize some generalized function here. However, I vaguely remember reading that its common to have to deal with an arbitrary scaling on the output of an fft, but don't quote me on that. I'll look into it and report back.

File: Sources/AudioKitUI/Visualizations/FFTView.swift (hunk 2 above)
Comment (Matt54) [reply Sources/AudioKitUI/Visualizations/FFTView.swift:52]: At the very least, you would think that there should be no need to have multiple lines to map the data appropriately even if an arbitrary scaling of the output data is required. I'll revisit this and see if I can sort out the magic numbers.

Comment (eljeff) [reply Sources/AudioKitUI/Visualizations/FFTView.swift:52]: Thanks for noting it. None of this is show stopping - just comments I can make from afar.
Again I really appreciate you doing this.

Comment (Matt54) [reply Sources/AudioKitUI/Visualizations/FFTView.swift:52]: I looked into this and I've cleared as much of the magic away as I could. Check out the following pull requests:
https://github.com/AudioKit/AudioKit/pull/2381
https://github.com/AudioKit/AudioKitUI/pull/7

Now, FFTView just has one map function which takes the expected amplitude values and maps them to a 0-1 range for the visualizer. The amplitude values matched what I was seeing in the spectrum analyzer in Ableton Live, so I'm content with it for now.

As far as FFTTap goes, it needed a scaling factor on it's output when it was not normalizing the fft data. Apple's documentation brings up this need for a scaling factor with their fft.

**Pull request merged**